### DISCUSSION TITLE
Request for test: new `Model::Find()` as a replacement for `Model()`

### DISCUSSION THREAD

**Discussion by apotonick** [trailblazer/trailblazer#257] [Ideas]
The `Model()` macro turned out to be quiet helpful, but super hard to extend, and it broke down as soon as you had the slightest change, e.g. finding the ID in `params[:song][:id]` and not in `params[:id]`. We now introduced `Model::Find()` and `Model::Build()` as replacements and are looking for testers, comments, suggestions etc. 

Here are the docs: https://model--ornate-heliotrope-1c15e2.netlify.app/2.1/docs/macro#macro-model-model-find

You need to add one line to your `Gemfile`.

```ruby
gem "trailblazer-macro", github: "trailblazer/trailblazer-macro", branch: "model"
```

Please test and find out
:ballot_box_with_check: if your code still works (original implementation of `Model()` hasn't changed)
:ballot_box_with_check: whether this is a helpful addition
:ballot_box_with_check: docs work for you

**Comment by seuros** [comment]
I propose the following syntax instead :

```ruby
step Model::Find(Song, params_key: :short_id ) # Song.find_by(id: params[:short_id]) 

step Model::Find(Song, key: :short_id ) # Song.find_by(short_id: params[:short_id]) 

step Model::Find(Song, params_key: [:product, id] ) # Song.find_by(id: params[:product][:id]) 

step Model::Find(Song, scopes: [:enabled, :copyrigthed] ) # Song.enabled.copyrigthed.find_by(id: params[:id]) 

step Model::Find(Song, scopes: ->(ctx, params:, **) { where(author: ctx['current_user'].id )} ) # Song.where(author: ctx['current_user'].id )}).find_by(id: params[:id]) 

step Model::Find(Song, scopes: :enabled ) # Song.enabled.find_by(id: params[:id]) 

step Model::Find(Song, method: :search_by) # Song.search_by(id: params[:id]) 

step Model::Find(Song, method: :find, key: nil) # Song.find(params[:id])

**Reply by apotonick** [reply]
:clap:  So, basically, you want to introduce `:method` instead of the hash style (which, indeed, might be confusing)?

**Reply by seuros** [reply]
Correct! it will be much easier to find and replace and to make defaults (method: :find_by).
Also the method could be a proc if someone want to do something more complex such as : 
`` method: (FeatureFlag?("use_elasticseach") ? :find_in_es_by : :find_by)``

**Reply by apotonick** [reply]
Very good point. :green_heart: I think the hash-style could be a sugaring on top, but maybe we can ditch it. Thing is, we will need another flag to figure out if `method` is to be called with keyword args or with positional (e.g. `find(id)`).

**Reply by apotonick** [reply]
After some more thoughts together with @seuros here's what we came up with.

```ruby
# explicit style:
      step Model::Find(Song, find_method: :find_method) # if it's _really_ Song.find_method(...)
      step Model::Find(Song, find_method: :find_by)
      step Model::Find(Song, find_method: :find_by, query_key: :id, params_key: :id)
      step Model::Find(Song, query: ->(ctx, params:, **) { where(id: params[:id_list]) })
# shorthand
      step Model::Find(Song, find_by: :id)
```

**Comment by apotonick** [comment]
Here are the new docs https://trailblazer.to/2.1/docs/macro/#macro-model-model-find please give us some input! :beers:

**Comment by apotonick** [comment]
https://dev.to/apotonick/fetching-models-using-the-new-modelfind-macro-jf3 :point_left: blog post!
